python_modules/dagster/dagster/_core/definitions/asset_automation_evaluator.py

```diff
@@ -0,0 +1,227 @@
+        )
+
+
+class OrAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_OrAutomationCondition", [("children", Sequence[AutomationCondition])]),
+):
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluations: List[ConditionEvaluation] = []
+        true_subset = context.empty_subset()
+        for child in self.children:
+            result = child.evaluate(context)
+            child_evaluations.append(result)
+            true_subset |= result.true_subset
+        return ConditionEvaluation(
+            condition=self, true_subset=true_subset, children=child_evaluations
+        )
+
+
+class InvertAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_InvertAutomationCondition", [("child", AutomationCondition)]),
+):
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluation = self.child.evaluate(context)
+        return ConditionEvaluation(
+            condition=self,
+            true_subset=context.candidate_subset - child_evaluation.true_subset,
+            children=[child_evaluation],
+        )
+
+
+class AssetAutomationEvaluator(NamedTuple):
+    """For now, this is an internal class that is used to help transition from the old format to the
+    new. Upstack, the original AutoMaterializePolicy class will be replaced with this.
+    """
+
+    condition: AutomationCondition
+    max_materializations_per_minute: Optional[int] = 1
+
+    def evaluate(
+        self, context: RuleEvaluationContext, from_legacy: bool
+    ) -> Tuple[
+        AutoMaterializeAssetEvaluation,
+        AssetDaemonAssetCursor,
+        AbstractSet[AssetKeyPartitionKey],
+    ]:
+        """Evaluates the auto materialize policy of a given asset.
+
+        Returns:
+        - An AutoMaterializeAssetEvaluation object representing serializable information about
+        this evaluation.
+        - The set of AssetKeyPartitionKeys that should be materialized.
+        - The set of AssetKeyPartitionKeys that should be discarded.
+        """
+        condition_evaluation = self.condition.evaluate(context)
+
+        # this is treated separately from other rules, for now
+        to_discard, discard_results = context.empty_subset(), []
+        if self.max_materializations_per_minute is not None:
+            discard_context = context.with_candidate_subset(condition_evaluation.true_subset)
+            condition = RuleCondition(
+                DiscardOnMaxMaterializationsExceededRule(limit=self.max_materializations_per_minute)
+            )
+            discard_condition_evaluation = condition.evaluate(discard_context)
+            to_discard = discard_condition_evaluation.true_subset
+            discard_results = discard_condition_evaluation.all_results
+
+        to_materialize = condition_evaluation.true_subset - to_discard
+
+        # hack to get the number of skipped asset partitions
+        num_skipped = 0
```

Comment:
This feels pretty brittle.

We'll need some form of this until we decide to stop supporting user code deployments with legacy policies, or we decide to stop showing skipped in the UI, right? I.e. we don't get to just eliminate it upstack?

I don't have any big ideas about how to make it less brittle, but a few small ideas:
- Rename the parameter to something like `report_num_skipped` and include docstring explaining that we can only report the number skipped if the condition follows the materialize rules / skip rules format.
- Make the assumptions about the structure of the rule more clear in code. E.g.
```python
skip_negation_condition = condition_evaluation.children[1]
skip_condition = skip_negation_condition.children[0]
```